README.md
<div align="center">
  <img width="500" src="./img/logo.png" alt="chakra + capsize">
  <br>
  <br>
  <br>
  <p>
    The power of <a href="https://seek-oss.github.io/capsize/">Capsize</a> inside <a href="https://chakra-ui.com/">Chakra UI</a>
  </p>
  <p>
    <sup>
      <a href="https://githubbox.com/ceteio/chakra-capsize/tree/main/example" target="_blank">Try it on Codesandbox</a>
    </sup>
  </p>
    <br>
  <br>
  <br>
</div>

```
yarn add @ceteio/chakra-capsize
```

**Usage**

```javascript
import { withCappedText } from "@ceteio/chakra-capsize/theme";
import { CappedText, CappedHeading } from "@ceteio/chakra-capsize";
import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import robotoFontMetrics from "@capsizecss/metrics/roboto";

const theme = extendTheme(
  defaultTheme,
  {
    fonts: {
      heading: "Roboto",
      body: "Roboto"
    },
    capHeights: {
      sm: 10,
      md: 14,
      lg: 18,
      xl: 24
    }
  },
  withCappedText({
    fontMetrics: {
      Roboto: robotoFontMetrics
    }
  })
);

const App = () => {
  return (
    <ChakraProvider theme={theme}>
      <div>
        <CappedHeading as="h1" size="2xl">
          Hi!
        </CappedHeading>
        <CappedText>
          This paragraph will have surrounding whitespace trimmed. It will also
          have space between the lines of text reduced to 0.
        </CappedText>
        <CappedText capHeight="lg" lineGap={4}>
          Setting a capHeight overrides any fontSize prop for more exact sizing.
          Meanwhile, a lineGap uses the Chakra 'spacings' scale to insert space
          between lines of text just like any other Chakra element.
        </CappedText>
      </div>
    </ChakraProvider>
  );
};
```

## Theming

Setting up your theme requires the following:

- A defined body & heading font
- Font metrics for each defined font
- Text size theme keys
- Extending the theme with `withCappedText`

### Body & Heading font

[Capsize recommends using non-system / non-fallback
fonts](https://seek-oss.github.io/capsize/) to get the best results & most
accurate trimming.

Specying fonts can be done in the following ways;

1. Using [chakra's fonts theme
   key](https://chakra-ui.com/docs/theming/theme#typography) _(default)_
   - `chakra-capsize` will use the theme's `fonts.body` &
     `fonts.heading` theme values for `<CappedText>` & `<CappedHeading>`
     respectively unless a more specific font family is passed in.
1. Setting `fontFamily` on `components.Text` / `components.Heading`.
   - This is particularly useful when you have variants that use different font
     faces such as pull quotes or buttons.
1. Passing `fontFamily` directly to a `<CappedText>` / `<CappedHeading>`.
   - The `fontFamily` prop passed will override any value on `Components.Text` /
     `components.Heading` or the fonts theme key.

### Font metrics for each defined font

Once your component has a font, `chakra-capsize` requires the corresponding font
metrics to be available for calculating the correct trim values.

Importantly, the key used when passing metrics to the
`withCappedText##fontMetrics` option must exactly match the font family name.

For example, when using the `Open Sans` font, you must set it up like so:

```javascript
withCappedText({
  fontMetrics: {
    'Open Sans': { ... }
  }
})

// ...

<CappedText fontFamily="Open Sans">...
```

### Text size theme keys

`chakra-capsize` has two primary configuration methods, mapping to the
underlaying implementation of `capsize`:

1. Text size
1. Space between lines of text

#### Text size

`chakra-capsize` allows defining the Text size as either a `fontSize` value, or
a `capHeight` value, where `capHeight` will override `fontSize` if set.

By default, `chakra-capsize` will use `fontSize` as it's the easiest to get
started with, and often more closely matches measurements from design tools such
as Figma & Sketch.

When you're ready to have more accurate control over the exact size of text, you
can use the `capHeight` prop by setting the `capHeights` key in your Chakra
theme:

```javascript
const theme = {
  capHeights: {
    sm: 10,
    md: 12,
    lg: 14
    // ...
  }
};
```

Then the corresponding values can be passed either directly to a `<CappedText>`
/ `<CappedHeading>` component:

```javascript
<CappedText capHeight="lg">...
```

or on variants, etc, in your component theme:

```javascript
const theme = {
  components: {
    Text: {
      variants: {
        huge: {
          capHeight: "2xl"
        }
      }
    }
  }
};
```

#### Space between lines

When multiple lines of text are displayed, a default gap of `0px` is applied,
resulting in lines of text whos capital letters press up againt each other.

This gap can be configured in multiple ways, two of which (`lineGap` &
`leading`) are closely related while the third (`lineHeight`) is available to
support values supplied in Chakra's default theme.

`lineGap` is the number of pixels between the bottom of one capital letter and
the top of a capital letter on the next line. `lineGap` uses the `spacings`
theme key.

`leading` can be thought of as `lineGap + capHeight`. Or, to put it another way;
the distance from the bottom of one capital letter to the bottom of a capital
letter on the next line. `leading` uses the `sizes` theme key.

`lineHeight` will set the value directly onto the text, however it's important
to note that it will _not_ affect the spacing at the top/bottom of the text
element (it will remain trimmed).

As with `capHeight`, `lineGap` & `leading` can be specified on the component
directly:

```javascript
<CappedText lineGap={6}>...
```

or on variants, etc, in your component theme:

```javascript
const theme = {
  components: {
    Text: {
      variants: {
        huge: {
          lineGap: 12
        }
      }
    }
  }
};
```

## `<CappedText>` & `<CappedHeading>`

Extending Chakra's [`<Text>`](https://chakra-ui.com/docs/typography/text) and
[`<Heading>`](https://chakra-ui.com/docs/typography/heading) components,
`<CappedText>` / `<CappedHeading>` uses `capsize` to trim space surrounding and
between lines of text.

Note when using `isTruncated` or `noOfLines` props: To avoid ascenders and
descenders from being cut off (due to `overflow: hidden`), an extra `<span>`
element is inserted inside the element when either `isTruncated` or `noOfLines`
is set.

### Props

All of `<Text>` / `<Heading>`'s props are fully supported. In addition,
the following props are available:

A note on responsive values: Each of the following props fully support
responsive formats (both array and object) with one restriction: a `base` size
_must_ be set, otherwise an error is thrown.

#### `capHeight`

Specifies the height, in pixels, of a capital letter in the line of text.
Internally, `chakra-capsize` trims surrounding whitespace via capsize which
makes `capHeight` equivalent to setting `min-height` on a `<div>`.

Can be a number, or one of the `capHeights` theme key (which resolves to a
number).

When set, `capHeight` will override any `fontSize` values set.

#### `lineGap`

Default: `0`.

When multiple lines of text are displayed, `lineGap` specifies the gap, in
pixels, between those lines.

`lineGap` uses the `spacings` theme scale.

When set, `lineGap` will override any `leading` or `lineHeight` values set.

#### `leading`

When multiple lines of text are displayed, `leading` specifies the distance
between one baseline to the next.

`leading` uses the `sizes` theme scale.

#### `sx`

Using the `sx` prop is fully supported, however behaviour is undefined when
attempting to set any of `fontSize`, `lineHeight`, `::before`, or `:;after`. It
is recommended to only set `fontSize` and `lineHeight` via props, and avoid
setting `::before` or `::after` at all.

#### `truncatedRef`

`<CappedText>` and `<CappedHeading>` can be passed a `ref` to access the React
DOM node under the hood. However, when either `isTruncated` or `noOfLines` is
set, an extra inner `<span>` is rendered; to access this inner DOM node, use the
`truncatedRef` prop.

## `withCappedText(options)`

```
import { withCappedText } from '@ceteio/chakra-capsize/theme'
```

Add support for capsize to your Chakra theme via `withCappedText`. It must be
passed in as an argument to Chakra's `extendTheme` to correctly extend the the
values required.

### `options.fontMetrics`

[Capsize](https://seek-oss.github.io/capsize/) requires a set of known font
metrics to correctly trim whitespace from around and between lines of text.

`chakra-capsize` also uses font metrics to support theming with Chakra.

The easiest way to generate font metrics is with the
[`@capsizecss/metrics`](https://github.com/seek-oss/capsize/tree/master/packages/metrics)
package:

```javascript
import robotoFontMetrics from "@capsizecss/metrics/roboto";
// ...
extendTheme(
  withCappedText({
    fontMetrics: {
      Roboto: robotoFontMetrics
    }
  })
);
```

_(See the [capsize docs](https://github.com/seek-oss/capsize#font-metrics) for
more ways of obtaining the necessary font metrics.)_

### `options.defaultFontSize`

_Default: `'md'`_

For `chakra-capsize` to work correctly out of the box, a default font size must
be set for all `<Text>` / `<CappedText>` / `<Heading>` / `<CappedHeading>`
components. The `defaultFontSize` option will use the `fontSizes` theme scale.

### `options.htmlFontSize`

_Default: `16`_

To safely convert between `rem` & `px` font sizes needed by Capsize,
`chakra-capsize` must know what the html font size is. All modern browsers set
this default to `16px`, and `chakra-capsize` does too.

You should only ever need to set this if you explicitly change the `font-size`
of the `html` element.

## Frequently Asked Questions

### Why are descenders cut off when using `isTruncated` or `noOfLines`?

To fix it, add a `lineGap` that is at least the height of the descenders.

This occurs due to `chakra-capsize`s default of `lineGap = 0` which instructs
capsize to set the container to be the same height as a capital letter. Then
when `isTruncated` sets `overflow: hidden`, the ascenders & descenders are cut
off. By setting a `lineGap` (or `leading` or `lineHeight`), the height of the
container is increased to stop the cut off while maintaining the correct


CHANGELOG.md
# @ceteio/chakra-capsize

## 1.2.0

### Minor Changes

- 350ff0f: Forward refs to the underlaying DOM nodes

### Patch Changes

- c1c94ae: Fix example font loading on codesandbox

## 1.1.0

### Minor Changes

- e590a56: Added an example project in `example/`

## 1.0.1

### Patch Changes

- ab7722e: Add repo URL to package.json for npm
- 598f34c: Correctly import React

## 1.0.0

### Major Changes

- 9a16269: Initial release

  ```
  yarn add @ceteio/chakra-capsize
  ```

  **Usage**

  ```javascript
  import { withCappedText } from "@ceteio/chakra-capsize/theme";
  import { CappedText, CappedHeading } from "@ceteio/chakra-capsize";
  import { ChakraProvider, extendTheme } from "@chakra-ui/react";
  import robotoFontMetrics from "@capsizecss/metrics/roboto";

  const theme = extendTheme(
    defaultTheme,
    {
      fonts: {
        heading: "Roboto",
        body: "Roboto"
      },
      capHeights: {
        sm: 10,
        md: 14,
        lg: 18,
        xl: 24
      }
    },
    withCappedText({
      fontMetrics: {
        Roboto: robotoFontMetrics
      }
    })
  );

  const App = () => {
    return (
      <ChakraProvider theme={theme}>
        <div>
          <CappedHeading as="h1" size="2xl">
            Hi!
          </CappedHeading>
          <CappedText>
            This paragraph will have surrounding whitespace trimmed. It will
            also have space between the lines of text reduced to 0.
          </CappedText>
          <CappedText capHeight="lg" lineGap={4}>
            Setting a capHeight overrides any fontSize prop for more exact
            sizing. Meanwhile, a lineGap uses the Chakra 'spacings' scale to
            insert space between lines of text just like any other Chakra
            element.
          </CappedText>
        </div>
      </ChakraProvider>
    );
  };
  ```


src/index.js
import React, { forwardRef } from "react";
import {
  Box,
  Text,
  Heading,
  useTheme,
  useToken,
  useStyleConfig
} from "@chakra-ui/react";
import { createStyleObject } from "@capsizecss/core";
import memoizeOne from "memoize-one";
import deepEqual from "lodash/isEqual";
import pickBy from "lodash/pickBy";
import omit from "lodash/omit";

// From https://github.com/chakra-ui/chakra-ui/blob/684012d868516412e91eb390379ccf7698536345/packages/utils/src/breakpoint.ts#L6
function analyzeCSSValue(value) {
  const num = parseFloat(value.toString());
  const unit = value.toString().replace(String(num), "");
  return { unitless: !unit, value: num, unit };
}

const capHeightToFontSize = ({ capHeight, fontMetrics }) =>
  capHeight / (fontMetrics.capHeight / fontMetrics.unitsPerEm);

const getBreakpointKeysArray = memoizeOne(theme =>
  Array.from(theme.__breakpoints.keys.values())
);

const getResponsiveValue = ({ theme, breakpoint, value }) => {
  const breakpointKeys = getBreakpointKeysArray(theme);

  for (let i = breakpointKeys.indexOf(breakpoint); i >= 0; i--) {
    if (
      // If there's a value set for the given breakpoint
      typeof value[breakpointKeys[i]] !== "undefined" &&
      // And that value isn't explicitly 'null'
      value[breakpointKeys[i]] !== null
    ) {
      // We've found our value
      return value[breakpointKeys[i]];
    }
  }
};

const detokenize = ({ theme, key, breakpoint, value }) => {
  if (value == null) {
    return undefined;
  }
  const concreteValue = getResponsiveValue({ theme, breakpoint, value });
  return concreteValue == null
    ? undefined
    : theme[key]?.[concreteValue] ?? concreteValue;
};

const responsivify = (theme, value) => {
  if (Array.isArray(value)) {
    return theme.__breakpoints.toObjectValue(value);
  }

  if (typeof value === "object" && theme.__breakpoints.isResponsive(value)) {
    return value;
  }

  if (value == null) {
    return undefined;
  }

  return { base: value };
};

const convertCssToPxValue = ({ theme, value, propName, breakpoint }) => {
  const valueMeta = analyzeCSSValue(value);
  let result;

  switch (valueMeta.unit) {
    case "rem":
      // Convert to a pixel-based size
      result = valueMeta.value * (theme.cappedText?.htmlFontSize ?? 16);
      break;
    case "px":
    case "":
      result = valueMeta.value;
      break;
    default:
      throw Error(
        `Unable to handle ${propName} value. Got ${JSON.stringify(
          value
        )}. Expected px or rem value.`
      );
  }

  return result;
};

const convertCssToLineheightValue = ({ capSizeConfig, fontMetrics }) => ({
  theme,
  value,
  propName,
  breakpoint
}) => {
  const valueMeta = analyzeCSSValue(value);
  let result;

  // A raw number or unitless string is a fontSize multiplier
  if (valueMeta.unitless) {
    if (capSizeConfig[breakpoint].capHeight) {
      // calculate the fontSize from the capHeight
      result =
        valueMeta.value *
        capHeightToFontSize({
          capHeight: capSizeConfig[breakpoint].capHeight,
          fontMetrics: fontMetrics[breakpoint]
        });
    } else {
      result = valueMeta.value * capSizeConfig[breakpoint].fontSize;
    }
  } else if (valueMeta.unit === "rem") {
    // A unit-ed number is assumed to be a rem value
    result = valueMeta.value * (theme.cappedText?.htmlFontSize ?? 16);
  } else {
    throw Error(
      `Unable to handle ${propName} value. Got ${JSON.stringify(
        value
      )}. Expected rem or unitless value.`
    );
  }

  return result;
};

const calculateResponsiveValues = ({
  theme,
  props,
  propName,
  themeKey,
  cssConverter = convertCssToPxValue
}) => {
  const responsiveValue = responsivify(theme, props[propName]);
  if (responsiveValue?.base == null) {
    throw new Error(
      `${propName} must specify a 'base' value when used responsively. Got ${JSON.stringify(
        props[propName]
      )}`
    );
  }

  let result = {};

  for (let breakpoint of Object.keys(responsiveValue)) {
    const concreteValue = detokenize({
      theme,
      key: themeKey,
      breakpoint,
      value: responsiveValue
    });

    result[breakpoint] = cssConverter({
      theme,
      value: concreteValue,
      propName,
      breakpoint
    });
  }

  return result;
};

const simplifyResponsiveValues = ({ theme, value, collapseBase = false }) => {
  // Convert to object syntax if not already
  const responsiveValue = responsivify(theme, value);

  const breakpointKeys = getBreakpointKeysArray(theme);
  let lastValueSet = responsiveValue[breakpointKeys[0]];
  let result = {
    [breakpointKeys[0]]: lastValueSet
  };

  for (let i = 1; i < breakpointKeys.length; i++) {
    // Skip breakpoints that aren't set
    if (typeof responsiveValue[breakpointKeys[i]] === "undefined") {
      continue;
    }
    const valueForBreakpoint = responsiveValue[breakpointKeys[i]];
    if (
      // Skip 'null' values (but not 'undefined')
      valueForBreakpoint !== null &&
      // We've hit a new value
      !deepEqual(valueForBreakpoint, lastValueSet)
    ) {
      lastValueSet = valueForBreakpoint;
      result[breakpointKeys[i]] = lastValueSet;
    }
  }

  if (
    collapseBase &&
    // When there's only a single item
    Object.keys(result).length === 1 &&
    // And that item is the 'base'
    typeof result[breakpointKeys[0]] !== "undefined"
  ) {
    // Just return that raw item on its own
    return result[breakpointKeys[0]];
  }

  return result;
};

const useCappedText = props => {
  const theme = useTheme();
  const { __breakpoints } = theme;

  let fontMetrics = {};
  let capSizeConfig = {};

  // ------------------
  // Font Family
  //
  // Parsing the `fontFamily` prop to figure out fontMetrics required by capSize
  // and our own calculations.
  // ------------------
  for (let breakpoint of __breakpoints.keys) {
    const responsiveFontFamily = responsivify(theme, props.fontFamily);
    const fontFamilyForBreakpoint = detokenize({
      theme,
      key: "fonts",
      breakpoint,
      value: responsiveFontFamily
    });

    // Determine the font family in use by splitting the css string
    // TODO: How do we handle comments, newlines, etc?
    const fontFamilies = fontFamilyForBreakpoint
      .split(",")
      .map(family => family.trim())
      .filter(Boolean);

    // Try to find the first font family we've got metrics for
    const fontFamily = fontFamilies.find(
      family => !!theme.cappedText.fontMetrics[family]
    );

    if (!fontFamily) {
      throw new Error(
        `Unable to determine font metrics for the given family: "${props.fontFamily}". Ensure the theme.cappedText.fontMetrics object is setup and the keys match the font families`
      );
    }

    fontMetrics[breakpoint] = theme.cappedText.fontMetrics[fontFamily];
  }

  // Initialise the config needed to pass to capSize for each breakpoint
  for (let breakpoint of __breakpoints.keys) {
    capSizeConfig[breakpoint] = {
      fontMetrics: fontMetrics[breakpoint]
    };
  }

  const useCapHeight = typeof props.capHeight !== "undefined";
  const useFontSize = !useCapHeight && typeof props.fontSize !== "undefined";

  const useLineGap = typeof props.lineGap !== "undefined";
  const useLeading = !useLineGap && typeof props.leading !== "undefined";
  const useLineHeight = !useLeading && typeof props.lineHeight !== "undefined";

  // ------------------
  // Cap Height
  //
  // Uses capHeight if set, otherwise fontSize
  // ------------------
  if (useCapHeight) {
    const responsiveCapHeights = calculateResponsiveValues({
      theme,
      props,
      propName: "capHeight",
      themeKey: "capHeights"
    });

    for (let [breakpoint, capHeight] of Object.entries(responsiveCapHeights)) {
      capSizeConfig[breakpoint].capHeight = capHeight;
    }
  } else if (useFontSize) {
    const responsiveFontSizes = calculateResponsiveValues({
      theme,
      props,
      propName: "fontSize",
      themeKey: "fontSizes"
    });

    for (let [breakpoint, fontSize] of Object.entries(responsiveFontSizes)) {
      capSizeConfig[breakpoint].fontSize = fontSize;
    }
  } else {
    // Shouldn't happen thanks to our defaultProp, but there's still a
    // case where `fontSize={null}`, etc.
    throw new Error(
      `One of fontSize or capHeight must be set on a cappedText component`
    );
  }

  // ------------------
  // Line Gap
  //
  // Uses lineGap if set, otherwise leading, otherwise lineHeight
  // ------------------
  if (useLineGap) {
    const responsiveLineGaps = calculateResponsiveValues({
      theme,
      props,
      propName: "lineGap",
      themeKey: "space"
    });

    for (let [breakpoint, lineGap] of Object.entries(responsiveLineGaps)) {
      capSizeConfig[breakpoint].lineGap = lineGap;
    }
  } else if (useLeading) {
    const responsiveLeadings = calculateResponsiveValues({
      theme,
      props,
      propName: "leading",
      themeKey: "sizes"
    });

    for (let [breakpoint, leading] of Object.entries(responsiveLeadings)) {
      capSizeConfig[breakpoint].leading = leading;
    }
  } else if (useLineHeight) {
    const responsiveLineHeights = calculateResponsiveValues({
      theme,
      props,
      propName: "lineHeight",
      themeKey: "lineHeights",
      cssConverter: convertCssToLineheightValue({ capSizeConfig, fontMetrics })
    });

    for (let [breakpoint, lineHeight] of Object.entries(
      responsiveLineHeights
    )) {
      capSizeConfig[breakpoint].leading = lineHeight;
    }
  } else {
    const breakpointKeys = getBreakpointKeysArray(theme);
    // Default to a zero lineGap, allowing the developer to insert their
    // own spacing as desired
    // Note we only set the 'base' value as it will propogate forward through
    // the responsive objects
    capSizeConfig[breakpointKeys[0]].lineGap = 0;
  }

  // ---------------
  // Filling in the blanks
  // ---------------
  // At this point, we have a responsive set of objects;
  // {
  //   base: {
  //     capHeight: 12,
  //     lineGap: 10
  //   },
  //   sm: {
  //     capHeight: 12,
  //   }
  //   md: {
  //     capHeight: 22
  //     lineGap: 14
  //   }
  // }
  //
  // Since there may be blanks, we want to make sure we fill them in (ie; no
  // explicit 'null's, and no missing fields.
  let prevValue = {};
  // NOTE: We iterate over __breakpoints.keys instead of
  // Object.keys(capSizeConfig) because order is important.
  for (let breakpoint of __breakpoints.keys) {
    if (typeof capSizeConfig[breakpoint] === "undefined") {
      continue;
    }

    capSizeConfig[breakpoint] = {
      ...prevValue,
      // Skip 'null' values, letting them be overridden by the previous
      // concrete value
      ...pickBy(capSizeConfig[breakpoint], value => value !== null)
    };
    prevValue = capSizeConfig[breakpoint];
  }

  // Then we simplify the object down to remove duplicates and save some work by
  // capSize.
  capSizeConfig = simplifyResponsiveValues({ theme, value: capSizeConfig });

  // ---------------
  // Running capSize
  // ---------------
  // Once we've filled in the blanks, we will use capSize to create the style
  // objects:
  // {
  //   base: {
  //     fontSize: 10,
  //     ::before: { content: '' }
  //   },
  //   sm: {
  //     fontSize: 10,
  //     ::before: { content: '' }
  //   }
  //   md: {
  //     fontSize: 15,
  //     ::before: { content: '' }
  //   }
  // }
  const capSizeStyleObjects = {};
  // Note: we don't care what order we do this step in, as long as every key is
  // hit. So we can use Object.keys(capSizeConfig)
  for (let breakpoint of Object.keys(capSizeConfig)) {
    // Generate the correct styles for the breakpoint
    capSizeStyleObjects[breakpoint] = createStyleObject(
      capSizeConfig[breakpoint]
    );
  }

  // ---------------
  // Responsive values
  // ---------------
  // Next, We want to change this into a set of values that are responsive:
  // {
  //   fontSize: {
  //     base: 10,
  //     sm: 10,
  //     md: 15
  //   },
  //   ::before: {
  //     content: {
  //       base: '',
  //       sm: '',
  //       md: ''
  //     }
  //   }
  // }
  //
  const result = {};

  for (let [breakpoint, capSizeStyleObject] of Object.entries(
    capSizeStyleObjects
  )) {
    // Then flatten it out into the result object
    for (let styleKey in capSizeStyleObject) {
      result[styleKey] = result[styleKey] || {};
      if (typeof capSizeStyleObject[styleKey] === "object") {
        // Pseudo selectos (`::before`, etc)
        for (let pseudoStyleKey in capSizeStyleObject[styleKey]) {
          result[styleKey][pseudoStyleKey] =
            result[styleKey][pseudoStyleKey] || {};
          result[styleKey][pseudoStyleKey][breakpoint] =
            capSizeStyleObject[styleKey][pseudoStyleKey];
        }
      } else {
        // scalar values (`fontSize`, etc)
        result[styleKey][breakpoint] = capSizeStyleObject[styleKey];
      }
    }
  }

  // ---------------
  // Simplifying things
  // ---------------
  // Then finally, we want to simplify those responsive objects down where
  // possible:
  // {
  //   fontSize: {
  //     base: 10,
  //     md: 15
  //   },
  //   ::before: { content: '' }
  // }
  // Do a last pass over the objects to simplify them. Can happen when multiple
  // breakpoint specify the same lineGap, but different capSize for example.
  for (let styleKey in result) {
    if (theme.__breakpoints.isResponsive(result[styleKey])) {
      // Scalar values (`fontSize`, etc)
      result[styleKey] = simplifyResponsiveValues({
        theme,
        value: result[styleKey],
        collapseBase: true
      });
    } else {
      // Pseudo selectors (`::before`, etc)
      for (let pseudoStyleKey in result[styleKey]) {
        result[styleKey][pseudoStyleKey] = simplifyResponsiveValues({
          theme,
          value: result[styleKey][pseudoStyleKey],
          collapseBase: true
        });
      }
    }
  }

  return result;
};

const CappedEl = forwardRef(
  ({ elName, Tag, fontFamilyType, truncatedRef, ...props }, ref) => {
    const { isTruncated, noOfLines, children } = props;
    const styleConfig = useStyleConfig(elName, props);
    const globalFontFamily = useToken("fonts", fontFamilyType);
    const defaultProps = useTheme().components[elName]?.defaultProps;
    const defaultFontFamily = defaultProps?.fontFamily ?? globalFontFamily;
    const cappedStyles = useCappedText({
      ...defaultProps,
      fontFamily: defaultFontFamily,
      ...styleConfig,
      ...props
    });

    const passthroughProps = omit(props, [
      "isTruncated",
      "noOfLines",
      "capHeight",
      "lineGap",
      "leading",
      "fontSize",
      "lineHeight"
    ]);

    return (
      <Tag
        ref={ref}
        {...passthroughProps}
        // Stop Chakra from looking up styles in the theme.components.Text object
        // since we've already done that above.
        // This allows us to pass custom styles while still leveraging the props
        // the <Text> component would normally accept.
        styleConfig={{}}
        // The theme styels first, then `sx` prop, then calculated cappedStyles
        sx={{ ...styleConfig, ...props.sx, ...cappedStyles }}
      >
        {isTruncated || noOfLines ? (
          // The inner <span> is necessary to avoid accidentally cutting off
          // ascenders / descenders when overflow: hidden is applied. In this way,
          // the correct capsize values are applied to the wrapping element which
          // is inherited by the <span>, but the overflow values are only applied
          // to the inner <span>.
          <Box
            ref={truncatedRef}
            as="span"
            {...{ isTruncated, noOfLines }}
            sx={isTruncated ? { display: "block" } : {}}
          >
            {children}
          </Box>
        ) : (
          children
        )}
      </Tag>
    );
  }
);

export const CappedText = forwardRef((props, ref) => (
  <CappedEl
    ref={ref}
    {...props}
    Tag={Text}
    elName="Text"
    fontFamilyType="body"
  />
));

export const CappedHeading = forwardRef((props, ref) => (
  <CappedEl
    ref={ref}
    {...props}
    Tag={Heading}
    elName="Heading"
    fontFamilyType="heading"
  />
));
